### DISCUSSION TITLE
Auto-append PR/Issue ID to all commits

### DISCUSSION THREAD

**Discussion by SyntevoAlex** [eclipse-platform/eclipse.platform.swt#583] [Ideas]
Summary of the problem:
* after moving to GitHub, many commits lack corresponding PR or Issue link.
* While GitHub allows to see which PR merged a specific commit, it's a bad idea to rely on that because SWT could move from GitHub some day.
* The suggestion is to introduce some scripts to auto-append the link when merging a PR.

Now a longer discussion:
* Before moving to GitHub, commits were expected to be named like `Bug 123456 - Some Description`. That is, there was always a bug number in the beginning.
* This makes it trivial to map a commit to Bugzilla entry and doesn't require any additional tools. Can be done easily in most environments.
* GitHub shows corresponding PR in commit description, for example see [8457de2](https://github.com/eclipse-platform/eclipse.platform.swt/commit/8457de260355f8e0f0dfec32bd6a3fed8858348d), the header on top of page says
  ```
  master #428
  v4960r3 ... I20221011-0640
  ```
  where `#428` is the corresponding PR.
  * However, not everyone knows that. I consider myself skilled, but I didn't know about this feature. What about newcomers? I would expect that many of them will not know this.
  * However, we can't be certain that SWT stays on GitHub forever. If SWT decides to move to any other platform, this mapping will be likely be lost, because it doesn't exist anywhere in the repo itself.

One solution will be to restore discipline and require good commit messages from committers 
* However, people are people, and that will require someone to spend time watching
  * This can be solved with GitHub-side check that doesn't allow to merge PR unless commits ave proper messages
* However, for PR-only commits (that is, without any Issue and just PR), the PR number can't be known in advance, so committer will have to create commit, create PR, then immediately adjust commit message to mention the PR identifier obtained.

The other solution would be to have GitHub-side scripts to auto-append required information in commit messages
* It will at least append links to PR/Issue involved
* It could also append various useful stuff like participating reviewers
* It could also maybe append the entire discussion from PR/Issue into the commit message? So that possible transition from GitHub will become really easy later

**Comment by laeubi** [comment]
>  after moving to GitHub, many commits lack corresponding PR or Issue link.
> ...
>  If SWT decides to move to any other platform, this mapping will be likely be lost, because it doesn't exist anywhere in the repo itself.

That is not true, each commit records its PR (and as we do not allow pushing directly each "new" commit is connected), you can see this in the web-ui and via standard git:

```
git config --add remote.origin.fetch +refs/pull/*/head:refs/remotes/origin/pull/*
git fetch origin
git describe --all  --contains <COMMIT>
```

I once suggested to ask JGit/EGit to add support for this in a more convenient way (like the fetch gtihub pull request menu item) but no one cared enough to do so ...

> However, not everyone knows that.

If you feel its missing to mention that in any commiter documentation, please add it, beside from that most things can be found by a websearch with `github <enter what ever you are curious about github>`, there is almaost everything answered already, [including this question](https://stackoverflow.com/questions/17818167/find-a-pull-request-on-github-where-a-commit-was-originally-created).

> The other solution would be

For me, writing a good commit message is enough, work with any tool, and many other projects seem to be fine with that as well, there is really no need to over complicate this:

1. Simple changes that do not require much discussion can be a PR and describe relevant information in the commit message
2. More complex ones can get an issue and you can simply copy the issue URL at the end of the commit + adding a short summary of the issue itself
